File: packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation_objc/FLTImageStreamHandler.m
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

@import Flutter;

#import "./include/camera_avfoundation/FLTImageStreamHandler.h"

@implementation FLTImageStreamHandler

- (instancetype)initWithCaptureSessionQueue:(dispatch_queue_t)captureSessionQueue {
  self = [super init];
  NSAssert(self, @"super init cannot be nil");
  _captureSessionQueue = captureSessionQueue;
  return self;
}

- (FlutterError *_Nullable)onCancelWithArguments:(id _Nullable)arguments {
  __weak typeof(self) weakSelf = self;
  dispatch_async(self.captureSessionQueue, ^{
    weakSelf.eventSink = nil;
  });
  return nil;
}

- (FlutterError *_Nullable)onListenWithArguments:(id _Nullable)arguments
                                       eventSink:(nonnull FlutterEventSink)events {
  __weak typeof(self) weakSelf = self;
  dispatch_async(self.captureSessionQueue, ^{
    weakSelf.eventSink = events;
  });
  return nil;
}
@end

File: packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTImageStreamHandler.swift
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

class FLTImageStreamHandler: NSObject, FlutterStreamHandler {

  /// The queue on which `eventSink` property should be accessed.
  let captureSessionQueue: DispatchQueue

  /// The event sink to stream camera events to Dart.
  ///
  /// The property should only be accessed on `captureSessionQueue`.
  /// The block itself should be invoked on the main queue.
  private(set) var eventSink: FlutterEventSink?

  init(captureSessionQueue: DispatchQueue) {
    self.captureSessionQueue = captureSessionQueue
    super.init()
  }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink)
    -> FlutterError?
  {
    captureSessionQueue.async { [weak self] in
      self?.eventSink = events
    }
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    captureSessionQueue.async { [weak self] in
      self?.eventSink = nil
    }
    return nil
  }
}
